### PR TITLE
PictureInPicture: replace discord's stream PiP 

### PR DESCRIPTION
**Pull request by lumap**
Say goodbye to [this](https://cdn.discordapp.com/attachments/1020339327831658629/1155203817852379298/image.png), and say hi to electron's PiP which can be moved outside of Discord. Can be turned off in the plugin's settings

### CODE REVIEW

File: src/plugins/pictureInPicture/README.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# PictureInPicture
+
+Adds a PiP button next to videos and replaces streams PiP
+
+![Example](https://i.imgur.com/HStdY53.png)
```
Comment (Vendicated) [inline src/plugins/pictureInPicture/README.md:5]: please do not use imgur, use github.

File: src/plugins/pictureInPicture/index.tsx [hunk 2]
```diff
@@ -33,6 +39,22 @@ export default definePlugin({
                 replace: "$&$1&&$2&&$self.renderPiPButton(),"
             },
         },
+        {
+            find: ".pictureInPictureVideo,",
+            predicate: () => settings.store.replaceStreams,
+            replacement: {
+                match: /onMouseLeave:\i,/,
```
Comment (Vendicated) [inline src/plugins/pictureInPicture/index.tsx:46]: this is way too vague, make it a bit more specific

File: src/plugins/pictureInPicture/index.tsx [hunk 3]
```diff
@@ -79,5 +101,27 @@ export default definePlugin({
...
+            mutations.forEach(mutation => {
+                const removedNode = (Array.from(mutation.removedNodes)[0]) as HTMLElement;
+                console.log(removedNode);
+                if (removedNode?.classList?.[0]?.startsWith("pictureInPictureWindow-") || removedNode?.classList?.[0]?.startsWith("video-") && removedNode.querySelector(".media-engine-video")) {
+                    document.exitPictureInPicture();
+                    observer.disconnect();
+                }
+            });
+        });
+        observer.observe(document, { childList: true, subtree: true });
```
Comment (Vendicated) [inline src/plugins/pictureInPicture/index.tsx:116]: this is very ugly, avoid dom manipulation / observing and do it properly, this isn't BD

File: src/plugins/pictureInPicture/README.md (hunk 1 above)
Comment (lumap) [reply src/plugins/pictureInPicture/README.md:5]: HStdY53

File: src/plugins/pictureInPicture/index.tsx [hunk 4]
```diff
@@ -33,6 +39,22 @@ export default definePlugin({
...
+            replacement: {
+                match: /className:.{0,10}pictureInPictureVideo/,
+                replace: "style:{display:\"none\"},$&"
+            }
+        },
+        {
+            find: ".pictureInPictureVideo,",
+            predicate: () => settings.store.replaceStreams,
+            replacement: {
+                match: /(\i)(=\i\.onJumpToChannel,.{0,500}"innerClassName".{0,4})return/,
```
Comment (Vendicated) [inline src/plugins/pictureInPicture/index.tsx:54]: what's the point of matching so much here if you're not gonna do anything with it? you can heavily simply this patch.
for example:
```js
match: /=(\i)\.renderBottomLeftControls,/,
replace: "=($self.enableStreamPiP($1),$1.renderBottomLeftControls),"
```
i didn't test it, but it probably works

File: src/plugins/pictureInPicture/index.tsx [hunk 5]
```diff
@@ -79,5 +101,19 @@ export default definePlugin({
                 )}
             </Tooltip>
         );
-    }, { noop: true })
+    }, { noop: true }),
+    enableStreamPiP: (jumpToChannel: any) => {
+        useEffect(() => {
+            const video = document.querySelector(".media-engine-video video") as HTMLVideoElement | null;
```
Comment (Vendicated) [inline src/plugins/pictureInPicture/index.tsx:107]: please use newlines

also don't have `{ code like this }`

**Comment by lumap** [comment]
cba to rewrite since rspack + not that useful

**Pull request closed without merging**